### ISSUE TITLE
Regression: Knative deployments not working in 2.2.1.Final

### ISSUE THREAD

**Issue by Sgitario** [quarkusio/quarkus#19950]
### Describe the bug

Deployments using Knative: https://quarkus.io/guides/deploying-to-kubernetes#knative, are failing with the following exception:

```
[INFO] [io.quarkus.kubernetes.deployment.KubernetesDeployer] Deploying to knative server: https://api.ocp48.dynamic.quarkus:6443/ in namespace: jose-test.
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:43 min
[INFO] Finished at: 2021-09-07T06:50:29+02:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal io.quarkus:quarkus-maven-plugin:2.2.1.Final:build (default) on project getting-started: Failed to build quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
[ERROR] 	[error]: Build step io.quarkus.kubernetes.deployment.KubernetesDeployer#deploy threw an exception: io.fabric8.kubernetes.client.KubernetesClientException: Cannot edit io.fabric8.knative.serving.v1.EditableService with visitors, no builder was found
[ERROR] 	at io.fabric8.kubernetes.client.ResourceHandlerImpl.edit(ResourceHandlerImpl.java:52)
[ERROR] 	at io.fabric8.kubernetes.client.dsl.internal.NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableImpl.acceptVisitors(NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableImpl.java:246)
[ERROR] 	at io.fabric8.kubernetes.client.dsl.internal.NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableImpl.inNamespace(NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableImpl.java:130)
[ERROR] 	at io.fabric8.kubernetes.client.dsl.internal.NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableImpl.inNamespace(NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableImpl.java:64)
[ERROR] 	at io.quarkus.kubernetes.deployment.KubernetesDeployer.lambda$deploy$4(KubernetesDeployer.java:179)
[ERROR] 	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
[ERROR] 	at io.quarkus.kubernetes.deployment.KubernetesDeployer.deploy(KubernetesDeployer.java:177)
[ERROR] 	at io.quarkus.kubernetes.deployment.KubernetesDeployer.deploy(KubernetesDeployer.java:102)
[ERROR] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[ERROR] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[ERROR] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[ERROR] 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
[ERROR] 	at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:820)
[ERROR] 	at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
[ERROR] 	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
[ERROR] 	at java.base/java.lang.Thread.run(Thread.java:829)
[ERROR] 	at org.jboss.threads.JBossThread.run(JBossThread.java:501)
```

This is reproduced for the Quarkus versions: 999-SNAPSHOT, 2.2.1.Final, 2.2.0.Final.
It works using Quarkus 2.1.0.Final.

### How to Reproduce?

1. Login to a OCP cluster with Knative installed (RH Serverless operator)
2. Create a new OCP project `oc new-project reproduce`
3. Generate the quarkus application: `mvn io.quarkus:quarkus-maven-plugin:2.2.1.Final:create -DprojectGroupId=org.acme -DprojectArtifactId=getting-started  -DclassName="org.acme.quickstart.GreetingResource" -Dpath="/hello" -Dextensions="resteasy,openshift"`
4. `mvn clean package -Dquarkus.kubernetes.deploy=true -Dquarkus.openshift.route.expose=true -Dquarkus.kubernetes-client.trust-certs=true -Dquarkus.kubernetes.deployment-target=knative`

**Comment by geoand** [comment]
@iocanel @manusa do you have any idea what's going on here? It looks like a Kubernetes Client issue.

**Comment by manusa** [comment]
It seems to be the same as https://github.com/fabric8io/kubernetes-client/issues/3433.

Originated in Quarkus after my PR #19795 which (for this case) removes the deprecated `deleteExisting()` DSL method call in combination with the K8s client version bump that contains changes to createOrReplace/edit/patch and other method internals.

We'll try to provide a fix upstream and include it in the 5.7.3 release that contains other fixes to the CRD generator that are needed in Quarkus 2.2.x. Is this OK? (/cc @metacosm @shawkins )

**Comment by geoand** [comment]
Thanks @manusa .

Can you also add this information to the JIRA for RHBQ I pinged you on?
Asking because this could affect the product delivery

**Comment by rupipal** [comment]
Following this guide for knative deployment ( https://quarkus.io/guides/deploying-to-kubernetes#knative ), which sets up a maven quickstart using the same Java class ( org.acme.quickstart.GreetingResource ), I'm facing exactly the same issue on minikube.

[ERROR] Failed to execute goal io.quarkus.platform:quarkus-maven-plugin:2.2.2.Final:build (default) on project kubernetes-quickstart: Failed to build quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
[ERROR] 	[error]: Build step io.quarkus.kubernetes.deployment.KubernetesDeployer#deploy threw an exception: io.fabric8.kubernetes.client.KubernetesClientException: Cannot edit io.fabric8.knative.serving.v1.EditableService with visitors, no builder was found
[ERROR] 	at io.fabric8.kubernetes.client.ResourceHandlerImpl.edit(ResourceHandlerImpl.java:52)
[ERROR] 	at io.fabric8.kubernetes.client.dsl.internal.NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableImpl.acceptVisitors(NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableImpl.java:246)
[ERROR] 	at io.fabric8.kubernetes.client.dsl.internal.NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableImpl.inNamespace(NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableImpl.java:130)
[ERROR] 	at io.fabric8.kubernetes.client.dsl.internal.NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableImpl.inNamespace(NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableImpl.java:64)
[ERROR] 	at io.quarkus.kubernetes.deployment.KubernetesDeployer.lambda$deploy$3(KubernetesDeployer.java:174)
[ERROR] 	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
[ERROR] 	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
[ERROR] 	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1655)
[ERROR] 	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
[ERROR] 	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
[ERROR] 	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
[ERROR] 	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
[ERROR] 	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
[ERROR] 	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
[ERROR] 	at io.quarkus.kubernetes.deployment.KubernetesDeployer.deploy(KubernetesDeployer.java:173)
[ERROR] 	at io.quarkus.kubernetes.deployment.KubernetesDeployer.deploy(KubernetesDeployer.java:103)
[ERROR] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[ERROR] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[ERROR] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[ERROR] 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
[ERROR] 	at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:820)
[ERROR] 	at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
[ERROR] 	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
[ERROR] 	at java.base/java.lang.Thread.run(Thread.java:829)
[ERROR] 	at org.jboss.threads.JBossThread.run(JBossThread.java:501)
[ERROR] -> [Help 1]

**Comment by geoand** [comment]
Fixed by #20016

**Comment by rupipal** [comment]
Thanks @geoand
